mumd/src/audio.rs

```diff
@@ -354,3 +366,16 @@ impl Audio {
         play_sounds.extend(samples.iter().skip(l));
     }
 }
+
+// moo
+fn get_resource(file: &str) -> Option<Cow<[u8]>> {
```

Comment:
You should include the lifetime associated with the `Cow`, in this case `'static`.

Reply:
Also, this function shouldn't return `None` when the file isn't found. Rather, it should return the default sound effect data wrapped in the `Cow`.

Reply:
I disagree. `get_resource` tries to open a file and return a Cow<[u8]> to it but it doesn't handle file not found which is up to the caller. It also makes it easier to move this later, should we ever need a) more kinds of byte files, or b) different behaviour for file not found depending on some other factor.

Come to think of it, it might be more clear if this returns `Result<&[u8], io::Error>`.

Reply:
Actually no, I'm happy with this as is. We need the Cow in order to return the slice as "owned" and None for if the file doesn't exist make sense to me.

Reply:
I think that a function called `get_resource` should always return some form of resource data. If its sole purpose is to read data from disk, I think it could get a name that is more indicative of the fact that it simply tries to read resource files from disk. If we want the function to be able to deal with different kinds of resources, we could make it generic. To make this generic over several types of resources, we could have a `trait Resource` and a `struct SoundEffect` for which `Resource` is implemented. Then, `get_resource` could have a type signature along the lines of
```rust
fn get_resource<R: Resource + Default>(path: &str) -> R
```
With reasonable methods on the `Resource` trait, this implementation wouldn't lose generality, and with the right `impl Deref` , it wouldn't be hard to work with the structs either.

Regarding point b, I don't think we'll want to both load resources in two different places in the code and have different behaviors on file not found.

If we decide that we do want to keep the current behavior of deferring handling of "file not found" to the caller, the function should return an `Option<Vec<u8>>`, since the `Cow` is only relevant when we might be dealing with either borrowed or owned data. `get_resource` only returns owned data and therefore, there is no reason to use a `Cow` when a `Vec` is sufficient. It would therefore be up to the caller to construct a `Cow` if it's necessary.